File: practice/algorithms/strings/common_child/common_child.hs
import qualified Data.Matrix as Matrix
import qualified Data.Map as Map
import Data.Maybe

insertVal s1 s2 row col m = Map.insert (row, col) v m where
    v = if s1 !! (row - 1) == (s2 !! (col - 1))
            then fromJust (Map.lookup (row - 1, col - 1) m) + 1
            else max (fromJust $ Map.lookup (row, col - 1) m) (fromJust $ Map.lookup (row - 1, col) m)

initMap n = Map.fromList ([((row, 0), 0) | row <- [0..n]] ++ [((0, col), 0) | col <- [0..n]])

constructTable s1 s2 n = foldl (\acc (row, col) -> insertVal s1 s2 row col acc) (initMap n) [(row, col) | row <- [1..n], col <- [1..n]]

main = do
    s1 <- getLine
    s2 <- getLine
    let n = length s1
        table = constructTable s1 s2 n
    print $ fromJust (Map.lookup (n, n) table)

File: practice/algorithms/strings/common_child/common_child.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <string>

using namespace std;

int main()
{
    string s1, s2;
    cin >> s1 >> s2;
    
    const size_t n = s1.length();
    
    vector<vector<int>> table(n + 1, vector<int>(n + 1, 0));
    for (size_t row = 1; row < n + 1; row++)
    {
        for (size_t col = 1; col < n + 1; col++)
        {
            if (s1[row - 1] != s2[col - 1])
            {
                table[row][col] = max(table[row][col - 1], table[row - 1][col]);
            }
            else
            {
                table[row][col] = table[row - 1][col - 1] + 1;
            }
        }
    }
    
    cout << table[n][n] << endl;
    
    return 0;
}

